packages/ckeditor5-ui/src/dialog/dialog.ts

```diff
@@ -0,0 +1,143 @@
+/**
+ * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module ui/dialog/dialog
+ */
+
+import View from '../view';
+import { type Editor, Plugin } from '@ckeditor/ckeditor5-core';
+import DialogView, { type DialogViewCloseEvent, DialogViewPosition } from './dialogview';
+import type { DialogActionButtonDefinition } from './dialogactionsview';
+
+import '../../theme/components/dialog/dialog.css';
+
+/**
+ * TODO
+ */
+export default class Dialog extends Plugin {
+	/**
+	 * TODO
+	 */
+	public readonly view: DialogView;
+
+	/**
+	 * TODO
+	 */
+	private _onHide: ( ( dialog: Dialog ) => void ) | undefined;
+
+	/**
+	 * @inheritDoc
+	 */
+	public static get pluginName() {
+		return 'Dialog' as const;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor: Editor ) {
+		super( editor );
+
+		this.view = new DialogView( editor.locale, () => {
+			return editor.editing.view.getDomRoot( editor.model.document.selection.anchor!.root.rootName )!;
+		}, () => {
+			return this.editor.ui.viewportOffset;
+		} );
+
+		this.view.on<DialogViewCloseEvent>( 'close', () => {
+			this.hide();
+		} );
+
+		editor.ui.view.body.add( this.view );
+		editor.ui.focusTracker.add( this.view.element! );
+	}
+
+	/**
+	 * TODO
+	 */
+	public show( {
```

Comment:
Different matter - one thing that I am missing in this whole dialog thing is a way to identify what dialog is/was opened. This is especially important if you want to plug in a custom behavior but only for certain dialogs. I propose introducing `id: string` property in `show()` and keep it in `Dialog` after a dialog was opened. It will be also passed to the events as one of parameters, so the integrator will be able to check it.

We could go even further and fire `show:<id>` and `hide:<id>` events.

Reply:
I'm not sure if ids will make the dialogs traceable (AFAIU id would be some random hash). If we want to make it usable for the integrators, maybe the dialog name/title would be better than id? We could use the `title` parameter passed with the `show()` method.

Still, I'm not entirely sure it's necessary. Custom behaviour can already be implemented within `onShow()`/\`onHide()\` callbacks while defining the dialog, so maybe let's discuss how it could be used.

Reply:
By id I meant a `string` that is set by a given feature that uses dialog. So more like a name for the dialog, rather than UID.

So, for example, F&R feature would call `dialog.show( { id: 'findAndReplace', ... }`.

Reply:
> Still, I'm not entirely sure it's necessary. Custom behaviour can already be implemented within `onShow()`/\`onHide()\` callbacks while defining the dialog, so maybe let's discuss how it could be used.

They cannot be implemented using `onShow` and `onHide` if you don't control the code that opens the dialog (and sets them). Which would be true for 3rd party customizations.